### ISSUE TITLE
trainers.segmentation JaccardLoss receiving num_classes, should be a List[int]?

### ISSUE THREAD

**Issue by robmarkcole** [microsoft/torchgeo#1896]
### Description

From [the trainer](https://github.com/microsoft/torchgeo/blob/8defbe4358b488f1eaa0712094c07d3f802f7523/torchgeo/trainers/segmentation.py#L113C1-L116C14):
```python
        elif loss == "jaccard":
            self.criterion = smp.losses.JaccardLoss(
                mode="multiclass", classes=self.hparams["num_classes"]
            )
```
`num_classes` is an int, Yet the docs for [JaccardLoss](https://smp.readthedocs.io/en/latest/losses.html#jaccardloss) state this is expected to be a _list of int_, presumably the indices of the output channels to use in the calc. Checking if this is a bug

### Steps to reproduce

NA

### Version

main

**Comment by calebrob6** [comment]
I verified the behavior of JaccardLoss here https://gist.github.com/calebrob6/658edaa59c68f0c0a510f8d9d7a41458.

Currently our code will always use the average jaccard over all classes regardless of what `self.hparams["num_classes"]` is.

We definitely should fix this.

@isaaccorley `ignore_index` support might be a cool feature for the torchseg fork.

**Comment by robmarkcole** [comment]
I suggest this requires a new arg of `classes_in_loss` which is a `List[int]`

**Comment by calebrob6** [comment]
Have you run into scenarios in which you would want to ignore multiple classes? If not, I think we can just pop ignore_index from the entire class list if it is given.

**Comment by robmarkcole** [comment]
I've not really used JaccardLoss, and commonly use [nn.CrossEntropyLoss](https://pytorch.org/docs/stable/generated/torch.nn.CrossEntropyLoss.html) where the `weights` parameter has been very effective for imbalanced datasets. If I understand the use of `classes` in JaccardLoss it has a simiar purpose, but is binary on/off per class, rather than a weighting

Coming back to your question, for multiclass datasets where I might only care about 1 or 2 classes, I can imagine ignoring multiple classes

**Comment by calebrob6** [comment]
+1 for fiddling with `weights` in CrossEntropyLoss.

Yep, I can imagine a few scenarios too (e.g. dropping loss over noisy classes). I would _usually_ do this by simply remapping all classes I didn't care about to a nodata mask value, then setting ignore_index to that value, but this is more elegant. My concern now would be implementing this functionality for the other loss functions.

**Comment by isaaccorley** [comment]
TorchSeg now supports both `classes` and `ignore_index` in `DiceLoss`, `JaccardLoss`, and `FocalLoss`.

**Comment by robmarkcole** [comment]
Thanks @isaaccorley

For anyone looking, they are on branch https://github.com/isaaccorley/torchseg/blob/losses/ignore-index/torchseg/losses/

**Comment by isaaccorley** [comment]
> I verified the behavior of JaccardLoss here https://gist.github.com/calebrob6/658edaa59c68f0c0a510f8d9d7a41458.
> 
> 
> 
> Currently our code will always use the average jaccard over all classes regardless of what `self.hparams["num_classes"]` is.
> 
> 
> 
> We definitely should fix this.
> 
> 
> 
> @isaaccorley `ignore_index` support might be a cool feature for the torchseg fork.

@robmarkcole it's merged into main now. I'll make another pre-release on pypi today.
